tcmur_cmd_handler.c

```diff
@@ -469,6 +474,216 @@ static int handle_read(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
 	return async_handle_cmd(dev, cmd, read_work_fn);
 }
 
+struct unmap_state {
+	uint16_t bddl;
+	uint8_t *par;
+};
+
+static int unmap_work_fn(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+{
+	struct tcmur_handler *rhandler = tcmu_get_runner_handler(dev);
+	struct unmap_state *unmap = cmd->cmdstate;
+	uint32_t block_size = tcmu_get_dev_block_size(dev);
+	uint64_t end_lba = tcmu_get_dev_num_lbas(dev) - 1;
+	struct iovec iovec;
+	size_t iov_cnt;
+	uint8_t *sense = cmd->sense_buf;
+	uint8_t *par;
+	uint16_t bddl = unmap->bddl, offset = 0;
+	void *iov_base;
+	int ret, i = 0;
+
+	iov_base = calloc(1, block_size);
+	iov_cnt = 1;
+
+	/* The first descriptor list offset is 8 in Data-Out buffer */
+	par = unmap->par + 8;
+
+	cmd->done = handle_generic_null_cbk;
+
+	while (bddl) {
+		uint64_t lba;
+		uint32_t nlbas;
+
+		lba = be64toh(*((uint64_t *)&par[offset]));
+		nlbas = be32toh(*((uint32_t *)&par[offset + 8]));
+
+		if (nlbas > VPD_MAX_UNMAP_LBA_COUNT) {
+			tcmu_err("Illegal parameter list LBA count %lu exceeds:%u\n",
+				 nlbas, VPD_MAX_UNMAP_LBA_COUNT);
+			ret = tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+						  ASC_INVALID_FIELD_IN_PARAMETER_LIST,
+						  NULL);
+			goto err;
+		}
+
+		if (lba + nlbas > end_lba || lba + nlbas < lba) {
+			tcmu_err("Illegal parameter list (lba + nlbas) %llu "
+				 "exceeds last lba %llu\n",
+				 lba + nlbas, end_lba);
+			ret = tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+						  ASC_LBA_OUT_OF_RANGE,
+						  NULL);
+			goto err;
+		}
+
+		tcmu_dev_dbg(dev, "Parameter list %d, lba: %llu, nlbas: %lu\n",
+			     i++, lba, nlbas);
+
+		/*
+		 * For now for the UNMAP emulator we will emulate the
+		 * fallocate() system call and will write zero to the
+		 * pre-allocated range.
+		 *
+		 * For example, for the file backstore it could pass
+		 * through the UNMAP command and implement its own
+		 * UNMAP callout by using fallocate() system call to
+		 * speed up this.
+		 */
```

Comment:
gluster has a unmap/discard command. qcow does too, but I think it is more complicated when supporting all the formats.

I think we want to end up where the SCSI command parsing like looping over param/descr lists is done here and calls a unmap callout for each chunk, because I do not think any one can do vectored/sg unmaps (they can just do the one start to end chunk in each call).

The code that does the write zeros based emulation below could then go into a lib function., so each handler could just call or set its unmap callout to the lib function initially. As we start to convert handlers to native unmap calls then they can just drop the lib call.

The write same code can then call the same callout if the unmap bit is set.